### PR TITLE
Annotate the public interface

### PR DESCRIPTION
**Pull request by layday**
This adds annotations to public attributes of the `Requirement`
class so that type checkers are not required ( :) to infer their type.

### CODE REVIEW

File: packaging/requirements.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 
 if TYPE_CHECKING:  # pragma: no cover
-    from typing import List
+    from typing import List, Optional as TOptional, Set
```
Comment (pradyunsg) [inline packaging/requirements.py:25]: Why not keep this as `Optional`?

Comment (layday) [reply packaging/requirements.py:25]: Because `Optional` is imported from pyparsing on line 11.

**Comment by layday** [comment]
Let me know if there's anything you'd like me to change.

**Comment by layday** [comment]
I would appreciate it if this simple fix was merged so I can stop nuking `packaging`'s `py.typed` in development to re-enable type inference ;)

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
@layday thanks!

**Pull request merged**